File: hogvm/python/stl.py
import time
from typing import Any, Optional
from collections.abc import Callable
import re
import json

from posthog.hogql.query import execute_hogql_query
from posthog.models import Team


def concat(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    def _to_concat_arg(arg) -> str:
        if arg is None:
            return ""
        if arg is True:
            return "true"
        if arg is False:
            return "false"
        return str(arg)

    return "".join([_to_concat_arg(arg) for arg in args])


def match(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return bool(re.search(re.compile(args[1]), args[0]))


def toString(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    if args[0] is True:
        return "true"
    elif args[0] is False:
        return "false"
    elif args[0] is None:
        return "null"
    else:
        return str(args[0])


def toInt(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    try:
        return int(args[0]) if name == "toInt" else float(args[0])
    except ValueError:
        return None


def ifNull(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    if args[0] is not None:
        return args[0]
    else:
        return args[1]


def length(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return len(args[0])


def empty(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return not bool(args[0])


def notEmpty(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return bool(args[0])


def lower(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return args[0].lower()


def upper(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return args[0].upper()


def reverse(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    return args[0][::-1]


def sleep(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    time.sleep(args[0])
    return None


def print(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int):
    if stdout is not None:
        stdout.append(f"{(' '.join(map(str, args)))}\n")
    return


def run(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int) -> list[Any]:
    if team is None:
        return []
    response = execute_hogql_query(query=args[0], team=team)
    return response.results


def jsonParse(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int) -> Any:
    return json.loads(args[0])


def jsonStringify(name: str, args: list[Any], team: Optional[Team], stdout: Optional[list[str]], timeout: int) -> str:
    if len(args) > 1 and isinstance(args[1], int) and args[1] > 0:
        return json.dumps(args[0], indent=args[1])
    return json.dumps(args[0])


STL: dict[str, Callable[[str, list[Any], Team | None, list[str] | None, int], Any]] = {
    "concat": concat,
    "match": match,
    "toString": toString,
    "toUUID": toString,
    "toInt": toInt,
    "toFloat": toInt,
    "ifNull": ifNull,
    "length": length,
    "empty": empty,
    "notEmpty": notEmpty,
    "lower": lower,
    "upper": upper,
    "reverse": reverse,
    "sleep": sleep,
    "print": print,
    "run": run,
    "jsonParse": jsonParse,
    "jsonStringify": jsonStringify,
}

File: hogvm/typescript/src/stl.ts
export const STL: Record<string, (args: any[], name: string, timeout: number) => any> = {
    concat: (args) => {
        return args.map((arg: any) => (arg === null ? '' : String(arg))).join('')
    },
    match: (args) => {
        const regex = new RegExp(args[1])
        return regex.test(args[0])
    },
    toString: (args: any[]) => {
        return String(args[0])
    },
    toUUID: (args) => {
        return String(args[0])
    },
    toInt: (args) => {
        return !isNaN(parseInt(args[0])) ? parseInt(args[0]) : null
    },
    toFloat: (args) => {
        return !isNaN(parseFloat(args[0])) ? parseFloat(args[0]) : null
    },
    ifNull: (args) => {
        return args[0] !== null ? args[0] : args[1]
    },
    length: (args) => {
        return args[0].length
    },
    empty: (args) => {
        return !args[0]
    },
    notEmpty: (args) => {
        return !!args[0]
    },
    lower: (args) => {
        return args[0].toLowerCase()
    },
    upper: (args) => {
        return args[0].toUpperCase()
    },
    reverse: (args) => {
        return args[0].split('').reverse().join('')
    },
    print: (args) => {
        // eslint-disable-next-line no-console
        console.log(...args)
    },
    jsonParse: (args) => {
        return JSON.parse(args[0])
    },
    jsonStringify: (args) => {
        // Recursively convert maps to objects
        function convert(x: any): any {
            if (x instanceof Map) {
                const obj: Record<string, any> = {}
                x.forEach((value, key) => {
                    obj[key] = convert(value)
                })
                return obj
            } else if (typeof x === 'object' && Array.isArray(x)) {
                return x.map(convert)
            } else if (typeof x === 'object' && x !== null) {
                const obj: Record<string, any> = {}
                for (const key in x) {
                    obj[key] = convert(x[key])
                }
                return obj
            }
            return x
        }
        if (args[1] && typeof args[1] === 'number' && args[1] > 0) {
            return JSON.stringify(convert(args[0]), null, args[1])
        }
        return JSON.stringify(convert(args[0]))
    },
}

export const ASYNC_STL: Record<string, (args: any[], name: string, timeout: number) => Promise<any>> = {
    sleep: async (args) => {
        await new Promise((resolve) => setTimeout(resolve, args[0] * 1000))
    },
}
